aten/src/ATen/native/native_functions.yaml

```diff
@@ -6508,6 +6508,12 @@
     SparseCPU: s_addmm_sparse_dense_cpu_
     SparseCUDA: s_addmm_sparse_dense_cuda_
 
+- func: _triton_bsr_dense_mm(Tensor bsr, Tensor dense) -> Tensor
```

Comment:
Why do you need this to be a native function? You can define it directly in python if you never need it here.

Reply:
Well, the we actually need it as it is being called inside `addmm`.

Reply:
If you call it as `at""_triton_bsr_dense_mm` you do indeed. But since you already have the op, you could also call it via `triton_kernel.call(...)` and that doesn't require a native function.
Being a native function just gives you all the codegen niceties (namespaces, out= variant, etc). But if you don't need any of these, you most likely don't want it to be native as it adds more constraints (BC guarantees in particular).

This is not a big deal, just wanted to mention it in case you actually want the flexibility.

Reply:
I'm not sure to follow your comment. How is it linked to linear and addmm implementations?

Reply:
Yeah, sorry, @albanD , was thinking about something else... Well, if we are fine with catching Exeptions, then we could give it a shot. I was just following the suite of similar triton native backends...

Reply:
I don't think it is worth slowing down merging this.
But if there are more needed, we might want to set up a chat with Brian to sit down and discuss what's the best way to integrate these functions.

Reply:
Might as well at least start the conversation here :) cc @bdhirsh

Reply:
If we want to have that discussion, it would help to clarify:
- Where do you want to call this function from? python? C++ library code? C++ native code?
- Where do you want to register impl for this function? C++ ATen? C++ library? python?

Reply:
And to point this out again, I am using the autogen out here. @albanD , if a python function is defined as, say, `f(a, b, out=None)`, it is safe to call it with `call(a, b, result)` in the CPP universe, right?

Reply:
Yes.

Reply:
Brilliant, than we could indeed remove it from `native`.

Reply:
> If we want to have that discussion, it would help to clarify:
> 
>     * Where do you want to call this function from? python? C++ library code? C++ native code?

All these possibilities.
> 
>     * Where do you want to register impl for this function? C++ ATen? C++ library? python?

Whichever path is the easiest to make sure the user does not have to do that explicitly.

Reply:
> Whichever path is the easiest to make sure the user does not have to do that explicitly.

How can this be whichever? Where is your kernel implemented? Is it implemented above or below autograd? Do you have C++ kernels by hand on top of the triton impl?

Reply:
Alban and I talked offline for a bit. It sounds like in cases like this (registering a triton kernel that you can call in C++), we'd like a convenient way to call the triton kernel from C++, but don't necessarily need all of the features that native_functions.yaml codegen gives you.

What do you think of this alternative?
- declare the triton op directly from python (under e.g. the `triton` namespace instead of `aten`
- From C++, directly invoke the op through the dispatcher


The boilerplate for that would look something like this:

python registration code:
```
        triton_lib = torch.library.Library("triton", "DEF")
        triton_lib.define("_triton_bsr_dense_mm(Tensor bsr, Tensor dense) -> Tensor")
        # In theory you could manually register an out= variant too,
        # although it'd probably be less work to just manually perform the copy_() in your C++ code
```

code to invoke from inside a C++ kernel:
```
    static auto triton_op = c10::Dispatcher::singleton()
                  .findSchemaOrThrow(
                      "triton::_triton_bsr_dense_mm", "")
                  .typed<Tensor(const Tensor&, const Tensor&)>()
    auto output = triton_op.call(tensorA, tensorB);
    out_tensor.copy_(output);
```

Reply:
Thank you, @bdhirsh , @albanD , I will give it a shot.

Reply:
@bdhirsh , whether the schema includes an out or not, I get the following failure:
```
================================================================================================================== FAILURES ==================================================================================================================
________________________________________________________________________________ TestSparseCSRCUDA.test_triton_bsr_dense_bmm_block_size_32_int32_cuda_float16 ________________________________________________________________________________
Traceback (most recent call last):
  File "/home/nik/git/Quansight/pytorch/test/test_sparse_csr.py", line 1527, in test_triton_bsr_dense_bmm
    res_tri = torch.ops.triton._triton_bsr_dense_mm_out(bsr, dense.transpose(-2, -1), out=res_tri_out)
  File "/home/nik/git/Quansight/pytorch/torch/_ops.py", line 527, in __call__  
    return self._op(*args, **kwargs or {})                                                                             
ValueError: vector::reserve                                                                                                                                                                                                                   ____________________________________
```
This issue is non-deterministic, and only manifests itself when several tests are being run in a group, say, with `pytest -sv test/test_sparse_csr.py -k triton`. I was not able to fail any individual test, however.

Reply:
@bdhirsh , I tried, albeit to the same result...

Reply:
@bdhirsh , unfortunately, same issue:
```
diff --git a/aten/src/ATen/native/sparse/SparseBlasImpl.cpp b/aten/src/ATen/native/sparse/SparseBlasImpl.cpp
index 1baa7cffe66..39fc7597a87 100644
--- a/aten/src/ATen/native/sparse/SparseBlasImpl.cpp
+++ b/aten/src/ATen/native/sparse/SparseBlasImpl.cpp
@@ -91,13 +91,13 @@ Tensor& _compressed_row_strided_mm_out(const Tensor& compressed, const Tensor& s
    && n % blocksize[0] == 0) {
     try {
       const auto triton_kernel = c10::Dispatcher::singleton()
-        .findSchemaOrThrow("triton::_triton_bsr_dense_mm_out", "")
-        .typed<Tensor&(const Tensor&, const Tensor&, Tensor&)>();
+        .findSchemaOrThrow("triton::_triton_bsr_dense_mm", "")
+        .typed<Tensor(const Tensor&, const Tensor&)>();
       // Call Triton only if dispatch key was overwritten.
       // This is not strictly necessary since the definition is done in Python,
       // but we leave it here for extra safety.
       if (triton_kernel.hasKernelForDispatchKey(c10::DispatchKey::SparseCsrCUDA)) {
-        return triton_kernel.call(compressed, strided, result);
+        return result.copy_(triton_kernel.call(compressed, strided));
       }
     } catch (const std::exception& e) {
       // The schema is not defined and/or the key is not overwritten,
diff --git a/test/test_sparse_csr.py b/test/test_sparse_csr.py
index 77c32175a9c..c900028d384 100644
--- a/test/test_sparse_csr.py
+++ b/test/test_sparse_csr.py
@@ -1481,13 +1481,13 @@ class TestSparseCSR(TestCase):
         kernel_invoked = [False]
 
         lib = torch.library.Library("triton", "DEF")
-        lib.define("_triton_bsr_dense_mm_out(Tensor bsr, Tensor dense, *, Tensor(a!) out) -> Tensor(a!)")
+        lib.define("_triton_bsr_dense_mm(Tensor bsr, Tensor dense) -> Tensor")
 
-        def impl(*args, **kwargs):
+        def impl(bsr, dense):
             kernel_invoked[0] = True
-            return bsr_dense_mm(*args, skip_checks=True, **kwargs)
+            return bsr_dense_mm(bsr, dense, skip_checks=True)
 
-        lib.impl("triton::_triton_bsr_dense_mm_out", impl, "SparseCsrCUDA")
+        lib.impl("triton::_triton_bsr_dense_mm", impl, "SparseCsrCUDA")
 
         # Note that each value in a non-zero block is in range block_size * [low^2, high^2).
         tensor = partial(make_tensor, device=device, dtype=dtype, low=0.5, high=1.5)
@@ -1523,9 +1523,7 @@ class TestSparseCSR(TestCase):
                 # Otherwise check correctness against bmm
                 # since nn.linear does not support bsr.dim() > 2.
                 res_dense = bsr.to_dense() @ dense.transpose(-2, -1)
-                res_tri_out = torch.empty_like(res_dense)
-                res_tri = torch.ops.triton._triton_bsr_dense_mm_out(bsr, dense.transpose(-2, -1), out=res_tri_out)
-                self.assertTrue(res_tri is res_tri_out)
+                res_tri = torch.ops.triton._triton_bsr_dense_mm(bsr, dense.transpose(-2, -1))
             self.assertEqual(res_tri, res_dense)

```

```
________________________________________________________________________________ TestSparseCSRCUDA.test_triton_bsr_dense_bmm_block_size_16_int64_cuda_float16 ________________________________________________________________________________
Traceback (most recent call last):                                                                                                                                                                                                            
  File "/home/nik/git/Quansight/pytorch/test/test_sparse_csr.py", line 1526, in test_triton_bsr_dense_bmm                                                                                                                                     
    res_tri = torch.ops.triton._triton_bsr_dense_mm(bsr, dense.transpose(-2, -1))                                                                                                                                                             
  File "/home/nik/git/Quansight/pytorch/torch/_ops.py", line 527, in __call__                                                                                                                                                                 
    return self._op(*args, **kwargs or {})                                                                             
ValueError: vector::reserve            
```
Note that it only happens when I call it in Python, the CPP part never fails.

Reply:
Shall we investigate that further, or do that in a follow-up?